{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import matplotlib\n",
    "import re"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare Files and softwares"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "set -e\n",
    "echo \"Downloading and installing blat...\"\n",
    "wget -q http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/blat/blat -O ./downloads/bin/blat\n",
    "chmod +x ./downloads/bin/blat\n",
    "\n",
    "echo \"Downloading and installing blast...\"\n",
    "wget -q https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools.static.binary -O ./downloads/bin/bedtools\n",
    "chmod +x ./downloads/bin/bedtools\n",
    "\n",
    "\n",
    "echo \"Downloading annotation files...\"\n",
    "pip install bedparse\n",
    "wget -q https://ftp.ensembl.org/pub/release-110/gtf/homo_sapiens/Homo_sapiens.GRCh38.110.gtf.gz -O ./downloads/Homo_sapiens.GRCh38.110.gtf.gz\n",
    "gunzip ./downloads/Homo_sapiens.GRCh38.110.gtf.gz\n",
    "bedparse gtf2bed ./downloads/Homo_sapiens.GRCh38.110.gtf > ./downloads/Ensembl_Homo_sapiens.GRCh38.110.bed\n",
    "wget -q https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz -O ./downloads/hg38.fa.gz\n",
    "gunzip ./downloads/hg38.fa.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def grep_bed_or_fasta(enst=None, mode=True, strand=None):\n",
    "    if mode:\n",
    "        enst = enst.split('.')[0]\n",
    "        proc = subprocess.Popen(\n",
    "        f\"grep {enst} ./downloads/Ensembl_Homo_sapiens.GRCh38.110.bed\", \n",
    "        stdin=None, \n",
    "        stdout=subprocess.PIPE, \n",
    "        stderr=subprocess.PIPE, \n",
    "        shell=True)\n",
    "        outinfo, errinfo = proc.communicate()\n",
    "        return outinfo\n",
    "    else:\n",
    "        cmd = \"./downloads/bin/bedtools getfasta -fi ./downloads/hg38.fa -bed ./downloads/tmp.bed > ./downloads/seq1.fa\" if strand=='+' \\\n",
    "            else \"./downloads/bin/bedtools getfasta -fi ./downloads/hg38.fa -bed ./downloads/tmp.bed -s > ./downloads/seq1.fa\"\n",
    "        proc = subprocess.Popen(\n",
    "        cmd, \n",
    "        stdin=None, \n",
    "        stdout=subprocess.PIPE, \n",
    "        stderr=subprocess.PIPE, \n",
    "        shell=True)\n",
    "        outinfo, errinfo = proc.communicate()\n",
    "        return None\n",
    "\n",
    "def getEnst(geneName1):\n",
    "    command = f\"\"\"grep -E 'transcript.*gene_name \"{geneName1}\"' ./downloads/Homo_sapiens.GRCh38.110.gtf | awk '$3 == \"exon\" {{split($0, a, \"; \"); for (i in a) if (match(a[i], /^exon_number \"[0-9]+\"/)) {{split(a[i], b, \" \"); if (b[2] > max) {{max=b[2]; line=$0}}}}}} END {{print line}}'\n",
    "    \"\"\"\n",
    "    proc = subprocess.Popen(\n",
    "        command, \n",
    "        stdin=None, \n",
    "        stdout=subprocess.PIPE, \n",
    "        stderr=subprocess.PIPE, \n",
    "        shell=True)\n",
    "    outinfo, errinfo = proc.communicate()\n",
    "    return outinfo\n",
    "\n",
    "\n",
    "def to_str(ori_list):\n",
    "    return [str(i) for i in ori_list]\n",
    "\n",
    "\n",
    "def Plot(start, end, chr_id):\n",
    "\n",
    "    start, end = start, end\n",
    "    SIZE_SCALE = abs(end - start) // 7\n",
    "    HALF_SIZE_SCALE = SIZE_SCALE // 2\n",
    "\n",
    "    fig = plt.figure(figsize=(16, 3))\n",
    "    ax = fig.add_subplot(111, aspect='equal')\n",
    "    plt.xlim([start, end])\n",
    "    plt.ylim([0, SIZE_SCALE])\n",
    "    ax.set_yticks([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "\n",
    "    xticks_start = (start//SIZE_SCALE + 1) * SIZE_SCALE\n",
    "    xticks_list = [i for i in range(xticks_start, end, SIZE_SCALE)]\n",
    "    ax.set_xticks(xticks_list)\n",
    "    tmp_list = [i//1000 for i in xticks_list]\n",
    "    ax.set_xticklabels([str(i//1000)+','+str(i%1000) for i in tmp_list], fontsize=10)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    def plot_tracks(chromStart, chromEnd, blockSizes, blockShift, thickStart, thickEnd, name,SIZE_SCALE, color, mode):\n",
    "        mid_line = SIZE_SCALE//4 if mode == \"ref\" else SIZE_SCALE*3//4\n",
    "        plt.plot([chromStart, chromEnd], [mid_line, mid_line], linestyle='-', color='black', alpha=0.7, lw=1, zorder=1)\n",
    "        mid_point = (chromStart + chromEnd) // 2\n",
    "        ARROW_SIZE = SIZE_SCALE//40\n",
    "        plt.plot([mid_point-ARROW_SIZE, mid_point+ARROW_SIZE], [mid_line-ARROW_SIZE*3, mid_line], linestyle='-', color='black', alpha=0.8, lw=0.8)\n",
    "        plt.plot([mid_point-ARROW_SIZE, mid_point+ARROW_SIZE], [mid_line+ARROW_SIZE*3, mid_line], linestyle='-', color='black', alpha=0.8, lw=0.8)\n",
    "        plt.text(chromEnd+ARROW_SIZE, mid_line-ARROW_SIZE, name, fontsize=10 , color='black')\n",
    "\n",
    "        def add_pathes(thickStart, thickEnd, chromStart, chromEnd, blockSizes, blockShift, SIZE_SCALE):\n",
    "            \n",
    "            for start, size in zip(blockShift, blockSizes):\n",
    "                if start + size <= thickStart-chromStart or start >= thickEnd-chromStart:\n",
    "                    rect = matplotlib.patches.Rectangle((chromStart + start, mid_line-SIZE_SCALE//20), size, SIZE_SCALE//10, color=color, alpha=1)\n",
    "                    ax.add_patch(rect)\n",
    "                elif start < thickStart-chromStart and start + size > thickStart-chromStart:\n",
    "                    truncated_blockSize = (thickStart-chromStart) - start\n",
    "                    rect = matplotlib.patches.Rectangle((chromStart + start, mid_line-SIZE_SCALE//20), truncated_blockSize, SIZE_SCALE//10, color=color, alpha=1)\n",
    "                    ax.add_patch(rect)\n",
    "                    new_start = chromStart + start + truncated_blockSize\n",
    "                    new_size = size - truncated_blockSize\n",
    "                    rect = matplotlib.patches.Rectangle((new_start, mid_line-SIZE_SCALE*3//20), new_size, SIZE_SCALE*3//10, color=color, alpha=1)\n",
    "                    ax.add_patch(rect)\n",
    "                elif start < thickEnd-chromStart and start + size > thickEnd-chromStart:\n",
    "                    truncated_blockSize = (start + size) - (thickEnd-chromStart)\n",
    "                    rect = matplotlib.patches.Rectangle((chromStart + start, mid_line-SIZE_SCALE*3//20), size - truncated_blockSize, SIZE_SCALE*3//10, color=color, alpha=1)\n",
    "                    ax.add_patch(rect)\n",
    "                    new_start = chromStart + start + size - truncated_blockSize\n",
    "                    rect = matplotlib.patches.Rectangle((new_start, mid_line-SIZE_SCALE//20), truncated_blockSize, SIZE_SCALE//10, color=color, alpha=1)\n",
    "                    ax.add_patch(rect)\n",
    "                else:\n",
    "                    rect = matplotlib.patches.Rectangle((chromStart + start, mid_line-SIZE_SCALE*3//20), size, SIZE_SCALE*3//10, color=color, alpha=1)\n",
    "                    ax.add_patch(rect)\n",
    "\n",
    "        add_pathes(thickStart, thickEnd, chromStart, chromEnd, blockSizes, blockShift, SIZE_SCALE)\n",
    "\n",
    "\n",
    "\n",
    "    def plot_from_bed(file, color, mode,SIZE_SCALE):\n",
    "        idx = 1\n",
    "        reference_color = {1:\"#482ff7\", 2:\"#086972\"}\n",
    "        old_start, old_end = 0, 0\n",
    "        with open(file) as f:\n",
    "            for q in f:\n",
    "                q = q.rstrip().split('\\t')\n",
    "                chromStart, chromEnd = int(q[1]), int(q[2])\n",
    "                name = q[3]\n",
    "                blockSizes = [int(i) for i in q[10].split(',')[:-1]]\n",
    "                blockShift = [int(i) for i in q[11].split(',')[:-1]]\n",
    "                thichStart, thickEnd = int(q[6]), int(q[7])\n",
    "                if mode == 'query':\n",
    "                    plot_tracks(chromStart, chromEnd, blockSizes, blockShift, thichStart, thickEnd, name, SIZE_SCALE, color=color, mode=mode)\n",
    "                else:\n",
    "                    if chromStart < old_end and old_end < chromEnd:\n",
    "                        idx += 1\n",
    "                    elif old_start > chromStart and old_start < chromEnd:\n",
    "                        idx += 1\n",
    "                    plot_tracks(chromStart, chromEnd, blockSizes, blockShift, thichStart, thickEnd, name, SIZE_SCALE, color=reference_color[idx], mode=mode)\n",
    "                    old_start, old_end = chromStart, chromEnd\n",
    "\n",
    "    plot_from_bed(\"./downloads/reference.bed\", color=\"#482ff7\", mode=\"ref\", SIZE_SCALE=SIZE_SCALE)\n",
    "    plot_from_bed(\"./downloads/query.bed\", color=\"#ff304f\", mode=\"query\", SIZE_SCALE=SIZE_SCALE)\n",
    "\n",
    "    ax.set_xlabel('Chromosome '+chr_id, fontsize=15, fontweight='bold')\n",
    "    ax.xaxis.set_label_coords(0.5, -0.2)\n",
    "\n",
    "    return fig\n",
    "\n",
    "def get_fig(enst1, enst2,info, seq, idx_fusion, primaryAligning):\n",
    "    # try:\n",
    "    if True:\n",
    "\n",
    "        outinfo_1 = grep_bed_or_fasta(enst1).decode('gbk')\n",
    "        outinfo_2 = grep_bed_or_fasta(enst2).decode('gbk')\n",
    "        with open(\"./downloads/reference.bed\", 'w') as w:\n",
    "            print(outinfo_1,end=\"\", file=w)\n",
    "            print(outinfo_2,end=\"\", file=w)\n",
    "\n",
    "        outinfo_1 = outinfo_1[:].split('\\n')[0].split(\"\\t\")\n",
    "        outinfo_2 = outinfo_2[:].split('\\n')[0].split(\"\\t\")\n",
    "        chr_id = outinfo_1[0]\n",
    "        strand = outinfo_1[5]\n",
    "        positions = [int(outinfo_1[1]), int(outinfo_1[2]), int(outinfo_2[1]), int(outinfo_2[2])]\n",
    "        start = min(positions) - 500\n",
    "        end = max(positions) + 500\n",
    "\n",
    "\n",
    "        with open(\"./downloads/tmp.bed\", 'w') as f:\n",
    "            print(\"\\t\".join([\"chr\"+chr_id, str(start), str(end), \"region_tmp\", \"0\", strand]), file=f)\n",
    "\n",
    "        grep_bed_or_fasta(mode=False, strand=strand)\n",
    "\n",
    "        with open(\"./downloads/seq2.fa\", 'w') as f:\n",
    "            print('>'+info+'\\n'+seq, end=\"\",file=f)\n",
    "\n",
    "        align_line = 6 if primaryAligning else 7\n",
    "        proc = subprocess.Popen(\n",
    "        f\"./downloads/bin/blat -t=dna -q=dna ./downloads/seq1.fa ./downloads/seq2.fa ./downloads/blat_out > /dev/null ; head -{align_line} ./downloads/blat_out | tail -1\", \n",
    "        stdin=None, \n",
    "        stdout=subprocess.PIPE, \n",
    "        stderr=subprocess.PIPE, \n",
    "        shell=True)\n",
    "        outinfo, errinfo = proc.communicate()\n",
    "\n",
    "        align_res = outinfo.decode('gbk').split('\\t')\n",
    "        '''\n",
    "        match\tmis- \trep. \tN's\tQ gap\tQ gap\tT gap\tT gap\tstrand\tQ        \tQ   \tQ    \tQ  \tT        \tT   \tT    \tT  \tblock\tblockSizes \tqStarts\t tStarts\n",
    "     \tmatch\tmatch\t   \tcount\tbases\tcount\tbases\t      \tname     \tsize\tstart\tend\tname     \tsize\tstart\tend\tcount\n",
    "        ---------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "        939\t4\t0\t0\t25\t48\t15\t7824\t+\tENST00000378842.1,GALT|ENST00000555981.5,IL11RA\t3067\t792\t1783\tchr9:34646362-34658983\t12621\t2458\t11225\t33\t29,6,16,22,6,43,16,19,38,15,27,11,17,46,69,12,15,19,46,17,16,53,55,18,5,22,59,27,36,19,31,24,89,\t792,824,831,848,870,877,920,938,958,997,1013,1043,1055,1073,1119,1189,1201,1219,1239,1291,1308,1325,1379,1435,1455,1462,1489,1550,1578,1614,1634,1670,1694,\t2458,2488,2494,2510,2635,2641,2685,2701,3048,3086,3101,3128,3139,3156,8855,8924,8937,9239,9258,10381,10399,10415,10468,10523,10541,10673,10698,10757,10937,11058,11077,11111,11136,\n",
    "\n",
    "        '''\n",
    "        blockSizes = [int(i) for i in align_res[-3].split(',')[:-1]] # 29,6,16,22,6,43,16,19,38,15,27,11,17,46,69,12,15,19,46,17,16,53,55,18,5,22,59,27,36,19,31,24,89,\t\n",
    "        Ref_pos = [int(i) for i in align_res[-1].split(',')[:-1]] # 2458,2488,2494,2510,2635,2641,2685,2701,3048,3086,3101,3128,3139,3156,8855,8924,8937,9239,9258,10381,10399,10415,10468,10523,10541,10673,10698,10757,10937,11058,11077,11111,11136,\n",
    "        Fatsq_pos = [int(i) for i in align_res[-2].split(',')[:-1]] # 792,824,831,848,870,877,920,938,958,997,1013,1043,1055,1073,1119,1189,1201,1219,1239,1291,1308,1325,1379,1435,1455,1462,1489,1550,1578,1614,1634,1670,1694,\n",
    "        \n",
    "        Ref_start, Ref_end, Fastq_start, Fastq_end = Ref_pos[0], Ref_pos[-1], Fatsq_pos[0], Fatsq_pos[-1]\n",
    "        last_exonSize = blockSizes[-1]\n",
    "        chromStart = start + Ref_pos[0] - Fatsq_pos[0] # 34646362 + 2458 - 792\n",
    "        exonStart  = start + Ref_pos[0] # chromStart + 792\n",
    "        exonEnd = start + Ref_pos[-1] + blockSizes[-1] # 34646362 + 11136 + 89\n",
    "        chromEnd = exonEnd + len(seq) - last_exonSize - Fastq_end # exonEnd + 3067 - 1783\n",
    "        blockSizes = [Fastq_start] + blockSizes[:] + [len(seq) - Fastq_end - last_exonSize]\n",
    "        Ref_pos = [0] + [i - Ref_start + Fastq_start for i in Ref_pos[:]]\n",
    "        Ref_end = Ref_pos[-1]\n",
    "        Ref_pos += [Ref_end + last_exonSize]\n",
    "    \n",
    "\n",
    "        blockSizes = \",\".join(to_str(blockSizes[:]))+\",\"\n",
    "        Ref_pos = \",\".join(to_str(Ref_pos[:]))+\",\"\n",
    "        query_bed = to_str([chr_id, chromStart, chromEnd, \"readthrough\"+str(idx_fusion), 0, strand, exonStart, exonEnd, 0, len(Fatsq_pos)+2, blockSizes, Ref_pos])\n",
    "\n",
    "        with open(\"./downloads/query.bed\", 'w') as w:\n",
    "            print(\"\\t\".join(query_bed), end=\"\", file=w)\n",
    "\n",
    "\n",
    "        final_start = start-1000 if start<chromStart else chromStart - 1000\n",
    "        final_end = end+1000 if end>chromEnd else chromEnd + 1000\n",
    "\n",
    "        fig = Plot(final_start, final_end, chr_id)\n",
    "        return fig\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`input using pasteboard`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_input = '''>ENST00000378842.1,GALT|ENST00000555981.5,IL11RA\n",
    "AGATTTTTCCAGCGGATCCCCCGGTGGCCTCATGTCGCGCAGTGGAACCGATCCTCAGCAACGCCAGCAGGCGTCAGAGGCGGACGCCGCAGCAGCAACCTTCCGGGCAAACGCCCACCAGCATATCCGCTACAACCCGCCTGCAGGATGAGTGGGTGCTGGTGTCAGCTCACCGCATGAAGCGGCCCTGGCAGGGTCAAGTGGAGCCCCAGCTTCTGAAGACAGTGCCACCGCCATGACCCTCTCAACCCTCTGTGTCCTGGGGCCATCCGAGCCAACGGAGAGGTGAATCCCCAGTACGATAGCACCTTCCTGTTTGACAACGACTTCCCCAGCTCTGCAGCCTGATGCCCCCAGTCCAGGACCCAGTGATCATCCGCCTTTTCCAAGCAAAGTCTGCTCGAGGAGTCTGTAAGGTCATTGTGCTTCCACCCCTGGTCGGATGTAACGCTGCCACTCATGTCGCGTCCGCTGAGATCCGGGCTGTTTGTTGATGCATGGGCCTCAGTCACAGAGGAGCTGGGTGCCCAGTACCCTTGGTGCAGATCTTTGAAAACAAAGGTGCCATGATGGGCTGTTCTAACCCCCACCCCCACTGGCCAGGTATGGGCCAGCAGTTTCCTGCCAGATATTGCCGCCAGCGTGAGGAGCGATCTCAGCAGGCCATAAGTGAGTCAGCATGGAGAGCCCCTGCTAATGGAGTACAGCCGCCAGGAGCTACTCAGGAAGGAACGTCTGCCGCCAACCAGTGAGCACTGGTTAGTACTGGTGTCCCCTTCTGGGGCAACATGCGCCCTACCAGACACTGCTGCTGCCCCGTCCGCGCATGTGGCGGCGGCTACCTGAGGCTGACCCCTGCTGAGCGTGATGATCTAGCCCTCCATCATGAAGAAGCTCTTGACCAAGTATGACAACCTCTTGAGACGTCCTTTCCCTCTACTCCATGGGCTGGCATGGCGGCTCCCACAGGATCAGAGGCTGGGGCCAACTGGAACCCATTGGCAGCTGCACGGCTCATTACTACCCTCCGCTCCTGCGCTTTTCTGCCACTGTCTCGGAAATTCATGGTTGGTCTACGAAATGCTTGCTCAGGCTCAGAGGGACCTCACCCCTGAGCAGATGAGCAGCAGCTGCTCAGGGCTGAGCAGGGTCCTGGTGGCCGTGGCTACAGCCCTGGTGTCTGCCTCCCTCCCCCTGCCCCAGGCCTGGGGCCCCCACCCAGGGGTCCAGTATGGGCAAGCCAGGCAGGTCCGTGAAGCTGTGTTGTCCTGGAGTGACTGCCGGGGACCCACAGTGTCCTGGTTTCGGATGGGGAGCCAAAGCTGGCTCCAGGGACCTGACTCTGGCTAGGGGCATGAACTGGTCCTGGCCCAGGCAGGACAGCACTGATGAGGGCACCTACATCTGCCAGACCCTGGATGGTGCACTTGGGGGGCACAGTGACCCTGCAGCTACGGGCTACCCCTCCAGCCCGCCCTGTTGTCCTCCTTGCCAAGCAGCCGACTATGAGAACTTCTCTTGCACTTGGAGTCCCAGCCAGATCAGCGGCGTTTACCCACCCGCTACCTCACCTCCTAACAGGAAGAAGACAGTCCTAGGAGCTGATAGCCAGAGGAGGAGTCCATCCACAGGGGCCCTGGCCATGCCCACAGGATCCCCTAGGGGGCTGCCCCGCTGTGTTGTCCACGGGGCTGGTTCTGGAGCCAGTACCGGATTAATGTGACTGAGGTGAACCCACTGGGTGCCAGCACACGCCTGCTGGATGTGAGCTTGCAGAGCATCTTGCGCCCTGACCCACACCCAGGGCCTGCCGGGTGAGAGTCCAGTACCAGGTTACCCCCGACGCCTGCGAGCCAGCTGGACATACCCTGCCTCCTGGCCCGTGCGAGCCCCCACTTCCTGCTCAAGTTCCGTTTGCAGTACCCGTCCGGCGCACCATCCAGCCTGGTCCACGGTGGAGCCAGCTGGACTAGGAGGAGGTGATCACAGATGCTGTGGCTGGGCTGCCCATGCTGTACGAGTCAGTGCCCGGGACTTTCTAGATGCTGGCACTGGAGCACCTGGAGCCCGGAGAGGCCGGGGAACTCCGAGCACTCGGGACCATACCAAAGGAGATACCAGCATGGGGCCTAGCTACACACGCCAGCCAGAGGTGGAAGCCTCAGGTGGACAGCCCTGCTCCTCCAAGGCCCTCCCTCCAACCACACCCTCGGCTACTTGATCACAGGGACTCTGTGGAGCAGGTAGCTGTGCTTGGCGTCTTTGGGAATCCTTTCTTTCCTGGGACTGGTGGCTGGGGCCCTGGCACTGCGGGCTCTGGTAAGTGACTGCCATTGGTCCCTCAGCCTCTGATCCTCACACATGCTCTGATGCCCATAGACCACATTCATCTCCACCCTTCATGACTGCCCGCTGAACCTGTCTGATTTGCTGGAACTACCTCCCCATACCTCCATCCCCCATGCCCCACTTGATTTTAACTGATTCCTCTCCTGACCCTTTACTAATAAACCCTTTGGCGGGGAGACTGAGATAACCCACATTGTTGGAGAGACAGCTGCCTTTCTATGCCCCAGGCTGAGGCTGAGACGGGGGGGAAGGATGGATCCCCAAAGCCTGGGTTCTTGGCCTCAGTGATTCCAGTGGACAGGCGTCCAGAGCTCCAAAACCTGTAGAGGACCCCAGGAGGGCTCGTTTCGGCAGATTCCACCTATAATTCTGTCTTGCTGGTGTGGATAGAAACCAGGCAGGACAGTAGTATCCCTATGGTTGGATCTCAGCTGGAAGTTCTGTTTGGAGCCCATTTCTTGTGAGACCCTGTATTTCAAATTTGCAGCTGAAAGGTGCTTGTAACTCTGATTTCACCCCAGAGTTGGAGTTCTGCTCAAGGAAACGTGTGTAATGTGTACTCTGTGTCCATGTGTGACCATGTGTCTGTGAGGCAGGGACATGTATTCTCTGCATGCATGTATGTAGGTGCCTGGGGAGTGTGTGTGGGTCCTTGGCTCTTGGCCTTTCCCCTTGCAGGGGTTGTTGCAGGTGTGAATAAAGAGAATAAGGAAGTTCT\n",
    ">ENST00000397832.7,IQCJ|ENST00000445224.6,SCHIP1\n",
    "CTCTTTGAGGCTTTCTGATGCCTAGAGTGAGCTCAACTCAAAGGTTTCCAGCCTCACACTCGCCTCACATTCCCCCACAGTCACATTGCGCTCTGTGATTCTGAGGAATACAGTGTGCCAGCATCCGATCCAGTCTCCTTTCACCTGCAGGTGTTCCAGAAACTTCAAAATGCGTCTGGAAGAACTGAAAAGATTGCAGAATCCTCTAGAACAAGTTAATGATGGAAAATATTCATTTGAAAACATTCAGCGAGCATGGCGAGAGTACCTGCAGCGGCAGGAGCCCCTGGGGAAGAGGAGCCCGTCCCCACCCTCTGTCTCCTCAGAGAAGCTGAGCAGCTCTGTCAGCATGAACACCTTCTCCGACAGCAGCACACCCGATTACCGAGAGGATGGGATGGATCTAGGCAGTGACGCCGGCAGCAGCAGCAGCAGCAGCCGCGCCAGTTCACAGTCCAACTCCACCAAAGTGACCCCTTGCTCCGAGTGCAAATCTTCATCGTCGCCGGGGGGCAGCCTGGACTTGGTGTCTGCCCTGGAGGACTATGAGGAGCCCTTCCCGGTCTACCAGAAGAAGGTGATTGATGAGTGGGCGCCGGAGGAGGACGGGGAGGAGGAGGAAGAGGAGGACGAGCGCGACCAGCGAGGGTACCGGGATGACCGCTCTCCGGCCCGGGAACCGGGGGACGTAAGCGCCAGGACCCGCAGCGGCGGCGGCGGGGGCAGGAGCGCCACCACCGCCATGCCGCCCCCGGTGCCCAACGGCAACCTCCACCAGCACGACCCCCAGGACCTCAGGCACAATGGCAACGTGGTGGTGGCTGGCCGGCCGAGCTGTTCCCGGGGCCCCCGCCGGGCGATCCAAAAGCCCCAGCCGGCTGGGGGCCGGCGCAGTGGCCGCGGCCCGGCGGCTGGGGGGCTCTGCCTTCAGCCCCCAGACGGCGGGACGTGCGTCCCCGAAGAGCCCCCGGTGCCACCTATGGATTGGGAGGCGCTGGAGAAGCATCTGGCCGGGCTGCAGTTCCGGGAGCAGGAGGTACGGAACCAGGGCCAGGCGAGGACCAACTCCACCTCCGCACAGAAAAATGAGAGAGAGTCTATCAGACAGAAGTTGGCACTTGGAAGCTTCTTTGATGATGGCCCAGGAATTTATACCAGCTGTAGCAAAAGTGGGAAGCCAAGCCTTTCCTCCCGACTGCAGAGTGGGATGAACTTGCAGATATGCTTTGTCAACGACAGTGGCAGTGATAAGGACAGTGATGCTGATGACAGTAAGACTGAAACCAGCTTGGACACCCCCTTGTCTCCCATGAGCAAACAGAGTTCTTCCTATTCTGATAGAGACACTACTGAAGAGGAGTCTGAATCCTTGGATGACATGGACTTCCTTACAAGGCAAAAGAAATTGCAAGCTGAAGCCAAAATGGCCCTTGCCATGGCCAAACCAATGGCCAAAATGCAAGTAGAAGTGGAGAAACAGAACAGGAAAAAGTCTCCCGTCGCTGATCTTCTGCCACACATGCCTCATATAAGTGAATGCTTGATGAAAAGAAGTTTAAAACCCACCGACCTGAGAGACATGACTATTGGGCAGCTACAAGTGATAGTCAATGATCTCCATTCCCAGATAGAAAGCTTGAATGAAGAGTTGGTCCAGCTGCTTCTCATCCGAGATGAGCTGCACACAGAGCAGGATGCCATGCTGGTGGACATTGAAGACTTGACCAGACATGCTGAAAGTCAGCAGAAGCACATGGCAGAGAAAATGCCTGCAAAGTGA\n",
    "'''\n",
    "# user_input = '''>tmpname,IQCJ,SCHIP1\n",
    "# CTCTTTGAGGCTTTCTGATGCCTAGAGTGAGCTCAACTCAAAGGTTTCCAGCCTCACACTCGCCTCACATTCCCCCACAGTCACATTGCGCTCTGTGATTCTGAGGAATACAGTGTGCCAGCATCCGATCCAGTCTCCTTTCACCTGCAGGTGTTCCAGAAACTTCAAAATGCGTCTGGAAGAACTGAAAAGATTGCAGAATCCTCTAGAACAAGTTAATGATGGAAAATATTCATTTGAAAACATTCAGCGAGCATGGCGAGAGTACCTGCAGCGGCAGGAGCCCCTGGGGAAGAGGAGCCCGTCCCCACCCTCTGTCTCCTCAGAGAAGCTGAGCAGCTCTGTCAGCATGAACACCTTCTCCGACAGCAGCACACCCGATTACCGAGAGGATGGGATGGATCTAGGCAGTGACGCCGGCAGCAGCAGCAGCAGCAGCCGCGCCAGTTCACAGTCCAACTCCACCAAAGTGACCCCTTGCTCCGAGTGCAAATCTTCATCGTCGCCGGGGGGCAGCCTGGACTTGGTGTCTGCCCTGGAGGACTATGAGGAGCCCTTCCCGGTCTACCAGAAGAAGGTGATTGATGAGTGGGCGCCGGAGGAGGACGGGGAGGAGGAGGAAGAGGAGGACGAGCGCGACCAGCGAGGGTACCGGGATGACCGCTCTCCGGCCCGGGAACCGGGGGACGTAAGCGCCAGGACCCGCAGCGGCGGCGGCGGGGGCAGGAGCGCCACCACCGCCATGCCGCCCCCGGTGCCCAACGGCAACCTCCACCAGCACGACCCCCAGGACCTCAGGCACAATGGCAACGTGGTGGTGGCTGGCCGGCCGAGCTGTTCCCGGGGCCCCCGCCGGGCGATCCAAAAGCCCCAGCCGGCTGGGGGCCGGCGCAGTGGCCGCGGCCCGGCGGCTGGGGGGCTCTGCCTTCAGCCCCCAGACGGCGGGACGTGCGTCCCCGAAGAGCCCCCGGTGCCACCTATGGATTGGGAGGCGCTGGAGAAGCATCTGGCCGGGCTGCAGTTCCGGGAGCAGGAGGTACGGAACCAGGGCCAGGCGAGGACCAACTCCACCTCCGCACAGAAAAATGAGAGAGAGTCTATCAGACAGAAGTTGGCACTTGGAAGCTTCTTTGATGATGGCCCAGGAATTTATACCAGCTGTAGCAAAAGTGGGAAGCCAAGCCTTTCCTCCCGACTGCAGAGTGGGATGAACTTGCAGATATGCTTTGTCAACGACAGTGGCAGTGATAAGGACAGTGATGCTGATGACAGTAAGACTGAAACCAGCTTGGACACCCCCTTGTCTCCCATGAGCAAACAGAGTTCTTCCTATTCTGATAGAGACACTACTGAAGAGGAGTCTGAATCCTTGGATGACATGGACTTCCTTACAAGGCAAAAGAAATTGCAAGCTGAAGCCAAAATGGCCCTTGCCATGGCCAAACCAATGGCCAAAATGCAAGTAGAAGTGGAGAAACAGAACAGGAAAAAGTCTCCCGTCGCTGATCTTCTGCCACACATGCCTCATATAAGTGAATGCTTGATGAAAAGAAGTTTAAAACCCACCGACCTGAGAGACATGACTATTGGGCAGCTACAAGTGATAGTCAATGATCTCCATTCCCAGATAGAAAGCTTGAATGAAGAGTTGGTCCAGCTGCTTCTCATCCGAGATGAGCTGCACACAGAGCAGGATGCCATGCTGGTGGACATTGAAGACTTGACCAGACATGCTGAAAGTCAGCAGAAGCACATGGCAGAGAAAATGCCTGCAAAGTGA\n",
    "# '''\n",
    "\n",
    "user_input = user_input.split('>')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`or load from file`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_input = []\n",
    "idx = -1\n",
    "with open(\"./downloads/example.fa\") as f:\n",
    "    for q in f:\n",
    "        if '>' == q[0]:\n",
    "            idx += 1\n",
    "            user_input.append(q.rstrip())\n",
    "        else:\n",
    "            user_input[idx] += ('\\n' + q.rstrip())\n",
    "        \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "primaryAligning = True\n",
    "\n",
    "idx = 0\n",
    "error_html = \"\"\n",
    "for record in user_input:\n",
    "    if record != '':\n",
    "        info, seq = \"\", \"\"\n",
    "        idx += 1\n",
    "        i = record.rstrip().split('\\n')\n",
    "        for k in i:\n",
    "            if ',' in k :\n",
    "                info += k\n",
    "            else:\n",
    "                seq += k\n",
    "\n",
    "        if 'ENST' in info:\n",
    "            enst1, enst2 = info.split('|')[0].split(',')[0].split('.')[0], info.split('|')[1].split(',')[0].split('.')[0]\n",
    "        else:\n",
    "            geneName1, geneName2 = info.split(',')[1], info.split(',')[2]\n",
    "            outinfo = getEnst(geneName1).decode('gbk')\n",
    "            matches = re.findall(r'transcript_id \"(.*?)\"', outinfo)\n",
    "            enst1 = matches[0]\n",
    "            outinfo = getEnst(geneName2).decode('gbk')\n",
    "            matches = re.findall(r'transcript_id \"(.*?)\"', outinfo)\n",
    "            enst2 = matches[0]\n",
    "        fig = get_fig(enst1, enst2, info, seq, idx, primaryAligning)\n",
    "\n",
    "        '''\n",
    "        select secondary alignment\n",
    "        '''\n",
    "        primaryAligning = False \n",
    "\n",
    "        fig.show()\n",
    "        \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
